dask/layers.py

```diff
@@ -97,6 +111,115 @@ def __init__(
         )
 
 
+class CreateRandomArrayDeps(CreateArrayDeps):
+    """Index-block info mapping for BlockwiseCreateRandomArray"""
+
+    def __init__(self, chunks: tuple, seeds: list, extra_chunks: tuple):
+        super().__init__(chunks + extra_chunks)
+        self.seeds = seeds
+        self.extra_chunks = extra_chunks
+
+    def __getitem__(self, idx: tuple):
+        block_info = super().__getitem__(idx)
+        # TODO should be able to do this with one accumulate call
+        if len(idx) == 0:
+            block_info["seed"] = self.seeds[0]
+        else:
+            n = 0
+            for dim, index in enumerate(idx[:-1]):
+                n += index * toolz.last(
+                    toolz.accumulate(operator.mul, self.num_chunks[(dim + 1) :])
+                )
+            n += idx[-1]
+            block_info["seed"] = self.seeds[n]
+        # TODO make this cleaner
+        if self.extra_chunks:
+            block_info["chunk-shape"] = block_info["chunk-shape"][
+                : -len(self.extra_chunks)
+            ]
+        return block_info
+
+    @classmethod
+    def __dask_distributed_pack__(cls, module: str, name: str, *args):
+        from distributed.protocol import serialize
+
+        chunks, seeds, extra_chunks = args
+        return (
+            module,
+            name,
+            chunks,
+            serialize(seeds),
```

Comment:
It may be easiest to separately serialize each seed, so that we can still do `block_info["seed"] = self.seeds[n]` in the getitem definition without needed any `__dask_distributed_unpack__` logic in this class.  However, we would then need to wrap the underlying creation function to handle deserialization of `block_info["seed"]`.

If we do use a function wrapper, it probably makes sense to fully move the `random_state_data` operation from the client to the worker (and avoid the need to ship the seeds through the scheduler altogether).  Do we know any RNG experts?  Can we simply shift the default/random seed by some factor related to the chunk index?

Reply:
Yeah, I haven't fully thought through the serialization story here. But at least, I think that `serialize` does drill down into the `list` and serialize the seeds separately, c.f. https://github.com/dask/distributed/blob/bef0308962345303123aba7ec6730757a61a4dfc/distributed/protocol/serialize.py#L263-L284 , and it shouldn't be necessary to deserialize them on the scheduler.

It would definitely be nice to be able to move a smaller representation of the random state to the workers, but at least this implementations doesn't ship *more* seed data than the current implementation.

Reply:
>But at least, I think that serialize does drill down into the list and serialize the seeds separately, c.f. https://github.com/dask/distributed/blob/bef0308962345303123aba7ec6730757a61a4dfc/distributed/protocol/serialize.py#L263-L284 , and it shouldn't be necessary to deserialize them on the scheduler.

Ah - You are correct that `serialize` will internally handle the elements of the seperately.  I hadn't considered the possibility of simply splitting the header/frames on the scheduler, but that may actually be a good way to go.

Regarding the question of moving random state generation to the workers:  I like the idea, but I also don't think you should worry about it in this PR.

Reply:
I was just experimenting with this.  It seems that we will need to add an explicit `iterate_collection=` argument to [`serialize`](https://github.com/dask/distributed/blob/bef0308962345303123aba7ec6730757a61a4dfc/distributed/protocol/serialize.py#L241).  Otherwise, we cannot guarentee that the scheduler will be able to access individual elements of a list/dict without needing to deserialize.  I can/will submit a PR to distributed to add this feature, but this means a proper `__dask_distributed_unpack__` soution here will not work with the latest release of distributed.

Reply:
> I can/will submit a PR to distributed to add this feature, but this means a proper __dask_distributed_unpack__ soution here will not work with the latest release of distributed.

On second thought, I don't think it makes sense to do this.  In general, is seems like a much cleaner solution to just call `serialize` seperately on each element on the client, rather adding an extra step of splitting the header/frames on the scheduler.

Reply:
> On second thought, I don't think it makes sense to do this. In general, it seems like a much cleaner solution to just call `serialize` seperately on each element on the client,

Yep, after taking another look at it, I agree with you

> rather than adding an extra step of splitting the header/frames on the scheduler.

Can you elaborate more on what you mean here? 

In general, I'm having a tough time reasoning about how deserialization should look on the worker. What would the idiomatic way of doing that look like? i.e.,
1. How does one identify whether one should deserialize? If it is coming back as `distributed.protocol.Serialized`, what would be the best way to identify that without relying on `distributed` being installed?
2. I ran into a couple of problems around making sure the numpy deserializer is registered on the worker. What am I doing wrong?

(I'm not demanding answers to the above, just what I'm thinking about :) )

Reply:
>>rather than adding an extra step of splitting the header/frames on the scheduler.
>
>Can you elaborate more on what you mean here?

If you call `serialize` on an entire collection, you will get a (header, frames) tuple, and so the scheduler will need to process/split the header and then convert this single tuple to a list of (header, frame) tuples.  The logic is not that complicated, but it is messier than simply serializing the individual elements to begin with.

>In general, I'm having a tough time reasoning about how deserialization should look on the worker. What would the idiomatic way of doing that look like? i.e.,

I have experimented with ways to do this, but I haven't established anything that I would call idiomatic yet...

>How does one identify whether one should deserialize? If it is coming back as distributed.protocol.Serialized, what would be the best way to identify that without relying on distributed being installed?
I ran into a couple of problems around making sure the numpy deserializer is registered on the worker. What am I doing wrong?
>
> 1. How does one identify whether one should deserialize? If it is coming back as distributed.protocol.Serialized, what would be the best way to identify that without relying on distributed being installed?

You have certainly isolated the crux of the problem here. It is easy to decide when to serialize (within an `__dask_distributed_pack__` definition).  However, it is not obvious when the worker will need to **de**serialize one or all of its arguments.  In my first pass, I was relying on pickle rather than serialize/deserialize, so the function simply needed to check if the argument(s) in question was a bytes object.  In [7415](https://github.com/dask/dask/pull/7415/files#r600887498), I just moved over to serialize/deserialize, and I am relying on  an explicit `"serialized"` label.  I don't "love" that solution, but it seems to work.

> 2. I ran into a couple of problems around making sure the numpy deserializer is registered on the worker. What am I doing wrong?

Not sure - but I'm hoping we wont need to do anything like this to check if the data is serialized.

Reply:
Yeah, I am probably doing something wrong, but I am finding that the *first* time I need to deserialize a numpy array on a worker it fails because that protocol hasn't been registered yet. Subsequent deserializations work fine.

I'm currently forcing the registration by relying on the side effect of [this](https://github.com/dask/dask/pull/7417/files#diff-bacbead89e1b64dbdf54cbe8878e9a52445945d2c2710b2a11edc65364c6b16dR37-R39) import, but I don't view that as a real solution. Perhaps @jrbourbeau has some insight as to what might be going wrong here?